snapshotter/demux/metrics/discovery/service_discovery.go

```diff
@@ -0,0 +1,103 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package discovery
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/containerd/containerd/log"
+	"github.com/firecracker-microvm/firecracker-containerd/snapshotter/demux/cache"
+)
+
+// ServiceDiscovery discovers metrics services in the snapshotter cache
+// and serves a []MetricsTarget on an HTTP server.
+type ServiceDiscovery struct {
+	// cache is the shared host-level cache of snapshotters also used by the demux snapshotter.
+	cache cache.Cache
+	// server is the HTTP server that returns a list of targets to pull metrics from and apply labels to those metrics.
+	server *http.Server
+}
+
+// NewServiceDiscovery returns a ServiceDiscovery with configured HTTP server and provided cache.
+func NewServiceDiscovery(port int, c cache.Cache) *ServiceDiscovery {
+	return &ServiceDiscovery{
+		server: &http.Server{
+			Addr: ":" + strconv.Itoa(port),
+		},
+		cache: c,
+	}
+}
+
+// Serve starts the HTTP server for receiving service discovery requests.
+func (sd *ServiceDiscovery) Serve() error {
+	sd.server.Handler = http.HandlerFunc(sd.serviceDiscoveryHandler)
+	err := sd.server.ListenAndServe()
+	if err != http.ErrServerClosed {
+		return err
+	}
+
+	return nil
+}
+
+// Shutdown shuts down the service discovery HTTP server.
+func (sd *ServiceDiscovery) Shutdown(ctx context.Context) error {
+	return sd.server.Shutdown(ctx)
+}
+
+// MetricsTarget represents the underlying JSON structure
+// that a Prometheus HTTP Service Discovery server is
+// expected to return.
+//
+// https://prometheus.io/docs/prometheus/latest/http_sd/
+type MetricsTarget struct {
+	Targets []string          `json:"targets"`
+	Labels  map[string]string `json:"labels"`
+}
+
+// serviceDiscoveryHandler scans the cache for snapshotters and their proxy server information,
+// and builds and returns a JSON response in the format of a Prometheus service discovery endpoint.
+func (sd *ServiceDiscovery) serviceDiscoveryHandler(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+
+	namespaces := sd.cache.List()
+	services := make([]MetricsTarget, 0)
```

Comment:
nit: I usually see the `[]MetricsTarget{}` syntax for empty slices.

Reply:
Not to further muddy the waters, but according to [this doc](https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices) from the Golang wiki about code reviews, the `var slicename []type` can be preferable in cases like this because the slice is an actual `nil` value instead of a non-nil, zero-length slice.  Since you are only `append`ing to the slice and not performing a direct assignment by index it works, because `append()` knows exactly what to do (the case is different with direct index `[n]` assignments).  You can test this behavior of `append()` like so:

```go
func main() {
	var services []string

	target := "test"

	services = append(services, target)

	for _, word := range(services) {
		fmt.Println(word)
	}
}
```

You'll get the expected result of `test`:
```
% ./main
test
```

Reply:
(I should add, this is also a nit and the code is functionally fine as-is, but if you do change it, consider the `var x []y` syntax as well.)

Reply:
looks like the linter doesn't like `var x[]y` syntax: 
https://github.com/firecracker-microvm/firecracker-containerd/runs/6103232280?check_suite_focus=true
> Error: snapshotter/demux/metrics/discovery/service_discovery.go:78:2: Consider preallocating `services` (prealloc)
var services []MetricsTarget

`services := []MetricsTarget{}` lints fine. marshalling a nil slice rather than an empty, non-nil slice such as `var slicename []type` to JSON would result in a null value rather than an empty JSON array. See https://apoorvam.github.io/blog/2017/golang-json-marshal-slice-as-empty-array-not-null/